### PR TITLE
306 concise mode for plan

### LINKED ISSUE #306: Concise mode for tofu plan
It would be great to have a way to shrink the output of terraform, which is very noisy. 
A `--short` or `--concise` parameter where only the resources updated would show, and all the refresh and noisy info on apply or plan would be filtered out.

### PR DESCRIPTION
**Pull request by SypherSP**
### Implemented -concise flag that reduces the noise in the output of plan

Instead of adding another ViewType like HumanView or JsonView, I chose to add a bool flag for concise within the global View that takes care of compact warnings and no colour so that in future if a concise mode needs to be implemented for other commands like apply or destroy, the same flag can be used. To implement that, a quality will need to be added to the respective commands. 

Adding a new ViewType would cause changes to files where there is no need for it to be known that the output is going to be in a concise format. Also View seems to be the right place to add this flag.

Resolves #306 

## Target Release

1.7.0

### CODE REVIEW

**Review by Yantrio — CHANGES_REQUESTED** [review]
Overall I understand the approach of not using a view, but I'm concerned that this now means we will have to litter the codebase with if `view.concise`. 

Would you be able to discuss why you took this approach in a bit more detail please?

Also, I notice that this PR only really has an impact on one small part of the logging, and I was assuming that after some conversation on #306 we could agree what would/could/should be logged by this consise view before we dove into code. This then would've helped make informed choices during the development phase.

Overall my personal opinion here is that this is attempting to be a quick fix, and instead we could've discussed how we would like the views to be handled in the long run and things like log levels, or flagging logs in certain ways so everywhere doesn't have to check if the view is concise or not.

**Comment by SypherSP** [comment]
Firstly, sorry that I didn't discuss my approach enough before moving on to the solution. 

My first thought was to add a new ViewType in arguments/types.go but I was constantly running into places like the following 
```
func NewOperation(vt arguments.ViewType, inAutomation bool, view *View) Operation {
	switch vt {
	case arguments.ViewHuman:
		return &OperationHuman{view: view, inAutomation: inAutomation}
	default:
		panic(fmt.Sprintf("unknown view type %v", vt))
	}
}
```
where I would have to add a case for arguments.ViewConcise. 
Screenshot 2024-02-06 at 10 46 14 AM

Also, it made more sense to have concise as a sub type of ViewHuman since that is where the intended use is. I faced a choice of how to implement that, thus posted [this](https://github.com/opentofu/opentofu/issues/306#issuecomment-1918360905) comment. 
Even in this method I could not figure out a simple way without appending a concise bool to multiple structs to get the concise argument where it was needed. 

I chose to go ahead with just two levels (concise or not) based on [this](https://github.com/opentofu/opentofu/issues/756#issuecomment-1915367253) comment on an issue suggesting log levels. 

Finally, my reason behind implementing it as an argument for view and not specific to the plan command, just like no-color or compact-warning are, was so that this concise flag can later be used along with other commands like apply or destroy. This seemed to me to be the least intrusive way and the most useable for future development of concise flag with other commands.

File: internal/command/views/hook_ui_test.go [hunk 1]
```diff
@@ -440,6 +440,39 @@ func TestPreRefresh(t *testing.T) {
...
+
+	action, err := h.PreRefresh(addr, states.CurrentGen, priorState)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	if action != tofu.HookActionContinue {
+		t.Fatalf("Expected hook to continue, given: %#v", action)
+	}
+	result := done(t)
```
[the remark this answers is not in the document]
Comment (SypherSP) [reply internal/command/views/hook_ui_test.go]: I have made the requested changes.

**Comment by Yantrio** [comment]
Sorry for the delay on the review and response for this one, Most of the team has been out on PTO and we're slowly working through the backlog of issues.

We really appreciate the contribution here, but after discussing it with some of the core contributors we had some concerns about the longevity of this implementation, and we would like to try and re-visit how we want to handle this in the future.

I will be raising a comment shortly on the issue so that we don't lose track. And I will link that here: <link incoming>

**Comment by SypherSP** [comment]
I understand that this might not be the end-to-end solution you were looking for and would love to help wherever I can when this actually gets implemented.

**Comment by Yantrio** [comment]
@SypherSP I'll be raising this issue in our triage meeting on Monday and we will get back with more information once that meeting has happened. Thank you for your patience on this one.

**Review by cube2222 — APPROVED** [review]
As James said, we've discussed this with the core team today. We've decided that including the flag definitely makes sense, but that we'd cut down the scope of this PR a bit.

The changes in the hook_ui view makes sense, everybody agrees that we'd like to hide the refreshing log lines in the concise view. The changes in the jsonformat/plan.go file aren't that obvious, and we'd like to keep them out for now. To me personally it seems like those things should still be printed in concise mode.

In practice we've also agreed that even though the presence of the flag (--concise) is a backwards-compatibility promise we'll be making, the way it affects the human output is not something that we're gonna make any promises on. In other words, this PR will introduce a baseline, and if community member want to expand it, they can create issues that we can then discuss.

I've left a few comments for the changes that we'd like to be rolled back for now. On that note, thanks for working on this, and sorry for the confusion. We've internally wrote down some learning on how we should better communicate such things in issues that we mark as accepted.

**Review by cube2222 — CHANGES_REQUESTED** [review]
Accidentally approved while meaning to request changes. See previous comments of mine.

File: internal/plans/quality.go [hunk 2]
```diff
@@ -17,4 +17,8 @@ const (
 	// semantically meaningful updates to state. They can sometimes still affect
 	// the format of state if applied.
 	NoChanges
+
+	// Concise is used to reduce the level of noise in the output and display
+	// only the important details.
+	Concise
```
[the remark this answers is not in the document]
Comment (SypherSP) [reply internal/plans/quality.go]: I've removed this and the reverted changes made to other related files.

File: internal/command/jsonformat/plan.go [hunk 3]
```diff
@@ -174,13 +174,12 @@ func (plan Plan) renderHuman(renderer Renderer, mode plans.Mode, opts ...plans.Q
...
 		}
 	}
-
 	if haveRefreshChanges {
 		renderer.Streams.Print(format.HorizontalRule(renderer.Colorize, renderer.Streams.Stdout.Columns()))
 		renderer.Streams.Println()
 	}
 
-	if willPrintResourceChanges {
+	if willPrintResourceChanges && !checkOpts(plans.Concise) {
```
[the remark this answers is not in the document]
Comment (SypherSP) [reply internal/command/jsonformat/plan.go]: I've made the requested change.

**Review by cube2222 — APPROVED** [review]
LGTM!

**Review by cam72cam — APPROVED** [review]

**Pull request merged**